patch/spinlock.patch

```diff
@@ -0,0 +1,12 @@
+--- ratelimiter.c	2021-01-19 01:34:52.624027083 +0100
```

Comment:
`peerlookup.patch` and `spinlock.patch` have two completely different functions.

`spinlock.patch` is currently only used for Virtual DSM. Its purpose is to solve the problem of kernel not defining the _raw_spin_unlock symbol https://github.com/runfalk/synology-wireguard/issues/66#issuecomment-753686408. I named it so because it solves a specific problem and, like `memneq.patch`, it can be manually applied via `APPLY_SPINLOCK_PATCH`.

`peerlookup.patch` is used to solve the Linux kernel oops problem that occurs when a connection is established. This problem occurs on DSM 7.0 and this is the only way I have been able to eliminate it. Maybe in stable version there will not be such problems and patch will be unnecessary.

Reply:
Aren't the diffs for `spinlock.patch` and `peerlookup.patch` identical though? They both add macros for the same missing spin lock symbols.

I get that `netlimiter.patch is something completely different though.

Reply:
They both add macros for spinlock, but for very different reasons.
`spinlock.patch` adds a macros for `spin_lock` and `spin_unlock` to resolve the issue of missing symbols on Virtual DSM.

`netlimiter.patch` defines macros for `spin_lock_bh` and `spin_unlock_bh`. These symbols are exported from the kernel, but an oops occurs when using WireGuard. This appears to be some kind of bug or unusual behavior in the DSM kernel.

```
2021-01-12T23:35:26+01:00  Workqueue: wg-kex-wg0 wg_packet_handshake_send_worker [wireguard]
2021-01-12T23:35:26+01:00  task: ffff88006bd0c4c0 ti: ffff88006bd80000 task.ti: ffff88006bd80000
2021-01-12T23:35:26+01:00  RIP: 0010:[<ffffffff8108fbc7>]  [<ffffffff8108fbc7>] queued_spin_lock_slowpath+0xc7/0x160
2021-01-12T23:35:26+01:00  RSP: 0018:ffff88006bd83c20  EFLAGS: 00010206
2021-01-12T23:35:26+01:00  RAX: 0000000f00016dee RBX: ffff88007985b2c8 RCX: ffff880070816dc0
2021-01-12T23:35:26+01:00  RDX: 0000000000040000 RSI: ffffffffffffffff RDI: ffff88006af50000
2021-01-12T23:35:26+01:00  RBP: ffff88006bd83c20 R08: 0000000000000101 R09: ffff88006bd83be0
2021-01-12T23:35:26+01:00  R10: 0000000000000000 R11: 051c3904092ea39d R12: ffff88006af40000
2021-01-12T23:35:26+01:00  R13: ffff88006af50000 R14: ffff88007985b2c8 R15: ffff88007985b2d0
2021-01-12T23:35:26+01:00  FS:  0000000000000000(0000) GS:ffff880070800000(0000) knlGS:0000000000000000
2021-01-12T23:35:26+01:00  CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
2021-01-12T23:35:27+01:00  CR2: 0000000f00016dee CR3: 000000000180a000 CR4: 00000000003406f0
2021-01-12T23:35:27+01:00  Stack:
2021-01-12T23:35:27+01:00   ffff88006bd83c30 ffffffff8155ba27 ffff88006bd83c70 ffffffffa0bd67da
2021-01-12T23:35:27+01:00   0000000000000000 ffff88006bd83d3c ffff88006bd83d94 ffff88007985b3f0
2021-01-12T23:35:27+01:00   ffff88007985b2c8 ffff88006bd83cc0 ffff88006bd83d28 ffffffffa0bcef86
2021-01-12T23:35:27+01:00  Call Trace:
2021-01-12T23:35:27+01:00   [<ffffffff8155ba27>] _raw_spin_lock_bh+0x27/0x30
2021-01-12T23:35:27+01:00   [<ffffffffa0bd67da>] wg_index_hashtable_insert+0x2a/0x1b0 [wireguard]
2021-01-12T23:35:27+01:00   [<ffffffffa0bcef86>] wg_noise_handshake_create_initiation+0x2f6/0x320 [wireguard]
2021-01-12T23:35:27+01:00   [<ffffffffa0bd2242>] wg_packet_send_handshake_initiation+0x82/0x100 [wireguard]
2021-01-12T23:35:27+01:00   [<ffffffff810da3df>] ? kprobe_flush_task+0x8f/0x130
2021-01-12T23:35:27+01:00   [<ffffffffa0bd2514>] wg_packet_handshake_send_worker+0x14/0x20 [wireguard]
2021-01-12T23:35:27+01:00   [<ffffffff810700ba>] worker_run_work+0x9a/0xe0
2021-01-12T23:35:27+01:00   [<ffffffffa0bd2500>] ? encrypt_packet+0x240/0x240 [wireguard]
2021-01-12T23:35:27+01:00   [<ffffffff810678ab>] process_one_work+0x1db/0x4e0
2021-01-12T23:35:27+01:00   [<ffffffff81067bdd>] worker_thread+0x2d/0x4a0
2021-01-12T23:35:27+01:00   [<ffffffff81067bb0>] ? process_one_work+0x4e0/0x4e0
2021-01-12T23:35:27+01:00   [<ffffffff8106c483>] kthread+0xd3/0xf0
2021-01-12T23:35:27+01:00   [<ffffffff8106c3b0>] ? kthread_worker_fn+0x160/0x160
2021-01-12T23:35:27+01:00   [<ffffffff8155c0af>] ret_from_fork+0x3f/0x80
2021-01-12T23:35:27+01:00   [<ffffffff8106c3b0>] ? kthread_worker_fn+0x160/0x160
```